Add RemoteTraceback for exception with multiple arguments

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51979 Add RemoteTraceback for exception with multiple arguments**

Fixes #25522 #30147 #34130 #44631 #48709 #51663

Taking reference of #40437 and [python](https://github.com/python/cpython/blob/e42b705188271da108de42b55d9344642170aa2b/Lib/concurrent/futures/process.py#L114-L131)

Due to lots of issues related to re-raising error, add a RemoteTraceback as the `__cause__` of the Exception in the main process.

When Python load object from different process using [pickle](https://docs.python.org/3/library/pickle.html#pickling-class-instances), it creates a new object with arguments. As the customized `Exception` is required for multiple arguments, which are not provided, it will causes the error.

```py
import torch.utils.data

class MyException(Exception):
    def __init__(self, msg, arg1, arg2):
        super().__init__(msg)

class MyDataset(torch.utils.data.IterableDataset):
    def __iter__(self):
        raise MyException("MyError", 1, 2)

if __name__ == '__main__':
    dl = torch.utils.data.DataLoader(MyDataset(), num_workers=1)
    list(iter(dl))
```

```
torch._utils._RemoteTraceback: Traceback (most recent call last):
  File "/Users/erjia/workspace/pytorch/torch/utils/data/_utils/worker.py", line 152, in _worker_loop
    fetcher = _DatasetKind.create_fetcher(dataset_kind, dataset, auto_collation, collate_fn, drop_last)
  File "/Users/erjia/workspace/pytorch/torch/utils/data/dataloader.py", line 52, in create_fetcher
    return _utils.fetch._IterableDatasetFetcher(dataset, auto_collation, collate_fn, drop_last)
  File "/Users/erjia/workspace/pytorch/torch/utils/data/_utils/fetch.py", line 21, in __init__
    self.dataset_iter = iter(dataset)
  File "/Users/erjia/workspace/test/test_exception.py", line 14, in __iter__
    raise MyException("MyError", 1, 2)
__mp_main__.MyException: MyError


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test/test_exception.py", line 19, in <module>
    list(iter(dl))
  File "/Users/erjia/workspace/pytorch/torch/utils/data/dataloader.py", line 518, in __next__
    data = self._next_data()
  File "/Users/erjia/workspace/pytorch/torch/utils/data/dataloader.py", line 1200, in _next_data
    return self._process_data(data)
  File "/Users/erjia/workspace/pytorch/torch/utils/data/dataloader.py", line 1226, in _process_data
    data.reraise()
  File "/Users/erjia/workspace/pytorch/torch/_utils.py", line 420, in reraise
    raise self.exc
Exception: Re-raise the exception in DataLoader worker process 0
```

ejguan committed:
Add RemoteTraceback for exception with multiple arguments

[ghstack-poisoned]

ssnl reviewed (changes requested):
You are not raising the original exception type (which is a requirement) as python does
https://github.com/python/cpython/blob/e42b705188271da108de42b55d9344642170aa2b/Lib/concurrent/futures/process.py#L120-L131

ejguan commented:
@SsnL 
IMO, the code you just mentioned only provided a way to embed traceback as a string into the `RemoteTraceback` as cause of remote `Exception` in the main process.
But, the main problem is how to reconstruct **the remote `Exception`** with multiple arguments in the main process, after getting it from child process by pickling. When pickling the `Exception`, `__reduce__` and `__setstate__` are used to save and recreate the `Exception` object like here.
https://github.com/python/cpython/blob/ed1007c0d74e658d1e6c9b51b12ce7501eb8cbf9/Objects/exceptions.c#L133-L164

As you can see, `__reduce__` returns a tuple of 2 or 3 elements. The first should be the Callable, in our case, it should be the third-party `Exception` class. And the second element is used for as the argument for the Callable (Constructor). https://docs.python.org/3/library/pickle.html#object.__reduce__

For a lot of `Exception` with multiple arguments, it uses `super().__init__(msg)` but it creates a problem for the `__reduce__` function as it only save the `msg` and drop the other arguments in the Derived `Exception`. So, in order to pickle these `Exception`, they should implement `__reduce__` by itself.
```py
>>> import pickle

>>> class MyException1(Exception):
>>>     def __init__(self, msg, arg1, arg2):
>>>         super().__init__(msg)
>>>         self.msg = msg
>>>         self.arg1 = arg1
>>>         self.arg2 = arg2

>>> class MyException2(Exception):
>>>     def __init__(self, msg, arg1, arg2):
>>>         super().__init__(msg)
>>>         self.msg = msg
>>>         self.arg1 = arg1
>>>         self.arg2 = arg2
...
>>>     def __reduce__(self):
>>>         return type(self), (self.msg, self.arg1, self.arg2)

>>> e1 = pickle.dumps(MyException1("Without reduce", 1, 2))
>>> print(pickle.loads(e1))  # Error
TypeError: __init__() missing 2 required positional arguments: 'arg1' and 'arg2'
>>> e2 = pickle.dumps(MyException2("With reduce", 1, 2))
>>> print(pickle.loads(e2))  # Works
With reduce
```
But, apparently it's not possible for us to inject the `__reduce__` function by ourself, as the construction signature is not known for these third-party `Exception`. The conclusion is the third-party `Exception` doesn't implement correctly, so it makes no sense to re-raise it from child to parent.
There are two proposals for these scenarios:
1 . We can definitely use original method to re-create the third-party `Exception`, which uses `msg` as the single argument for construction. But, for the multiple-argument `Exception`, I propose to create a `LocalException` with the cause of `RemoteTraceback`. (Maybe also mention third-party `Exception` is incorrectly created)
2. In order to keep everything consistent for all `Exception`, use the `LocalException` as the PR suggested.

Any suggestion on it?

ejguan committed:
Update on "Add RemoteTraceback for exception with multiple arguments"


Fixes #25522 #30147 #34130 #44631 #48709 #51663

Taking reference of #40437 and [python](https://github.com/python/cpython/blob/e42b705188271da108de42b55d9344642170aa2b/Lib/concurrent/futures/process.py#L114-L131)

Due to lots of issues related to re-raising error, add a RemoteTraceback as the `__cause__` of the Exception in the main process.

When Python load object from different process using [pickle](https://docs.python.org/3/library/pickle.html#pickling-class-instances), it creates a new object with arguments. As the customized `Exception` is required for multiple arguments, which are not provided, it will causes the error.

```py
import torch.utils.data

class MyException(Exception):
    def __init__(self, msg, arg1, arg2):
        super().__init__(msg)

class MyDataset(torch.utils.data.IterableDataset):
    def __iter__(self):
        raise MyException("MyError", 1, 2)

if __name__ == '__main__':
    dl = torch.utils.data.DataLoader(MyDataset(), num_workers=1)
    list(iter(dl))
```

```
torch._utils._RemoteTraceback: Traceback (most recent call last):
  File "/Users/erjia/workspace/pytorch/torch/utils/data/_utils/worker.py", line 152, in _worker_loop
    fetcher = _DatasetKind.create_fetcher(dataset_kind, dataset, auto_collation, collate_fn, drop_last)
  File "/Users/erjia/workspace/pytorch/torch/utils/data/dataloader.py", line 52, in create_fetcher
    return _utils.fetch._IterableDatasetFetcher(dataset, auto_collation, collate_fn, drop_last)
  File "/Users/erjia/workspace/pytorch/torch/utils/data/_utils/fetch.py", line 21, in __init__
    self.dataset_iter = iter(dataset)
  File "/Users/erjia/workspace/test/test_exception.py", line 14, in __iter__
    raise MyException("MyError", 1, 2)
__mp_main__.MyException: MyError


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test/test_exception.py", line 19, in <module>
    list(iter(dl))
  File "/Users/erjia/workspace/pytorch/torch/utils/data/dataloader.py", line 518, in __next__
    data = self._next_data()
  File "/Users/erjia/workspace/pytorch/torch/utils/data/dataloader.py", line 1200, in _next_data
    return self._process_data(data)
  File "/Users/erjia/workspace/pytorch/torch/utils/data/dataloader.py", line 1226, in _process_data
    data.reraise()
  File "/Users/erjia/workspace/pytorch/torch/_utils.py", line 420, in reraise
    raise self.exc
Exception: Re-raise the exception in DataLoader worker process 0
```

[ghstack-poisoned]

ejguan committed:
Update on "Add RemoteTraceback for exception with multiple arguments"


Fixes #25522 #30147 #34130 #44631 #48709 #51663

Taking reference of #40437 and [python](https://github.com/python/cpython/blob/e42b705188271da108de42b55d9344642170aa2b/Lib/concurrent/futures/process.py#L114-L131)

Due to lots of issues related to re-raising error, add a RemoteTraceback as the `__cause__` of the Exception in the main process.

When Python load object from different process using [pickle](https://docs.python.org/3/library/pickle.html#pickling-class-instances), it creates a new object with arguments. As the customized `Exception` is required for multiple arguments, which are not provided, it will causes the error.

```py
import torch.utils.data

class MyException(Exception):
    def __init__(self, msg, arg1, arg2):
        super().__init__(msg)

class MyDataset(torch.utils.data.IterableDataset):
    def __iter__(self):
        raise MyException("MyError", 1, 2)

if __name__ == '__main__':
    dl = torch.utils.data.DataLoader(MyDataset(), num_workers=1)
    list(iter(dl))
```

```
torch._utils._RemoteTraceback: Traceback (most recent call last):
  File "/Users/erjia/workspace/pytorch/torch/utils/data/_utils/worker.py", line 152, in _worker_loop
    fetcher = _DatasetKind.create_fetcher(dataset_kind, dataset, auto_collation, collate_fn, drop_last)
  File "/Users/erjia/workspace/pytorch/torch/utils/data/dataloader.py", line 52, in create_fetcher
    return _utils.fetch._IterableDatasetFetcher(dataset, auto_collation, collate_fn, drop_last)
  File "/Users/erjia/workspace/pytorch/torch/utils/data/_utils/fetch.py", line 21, in __init__
    self.dataset_iter = iter(dataset)
  File "/Users/erjia/workspace/test/test_exception.py", line 14, in __iter__
    raise MyException("MyError", 1, 2)
__mp_main__.MyException: MyError


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test/test_exception.py", line 19, in <module>
    list(iter(dl))
  File "/Users/erjia/workspace/pytorch/torch/utils/data/dataloader.py", line 518, in __next__
    data = self._next_data()
  File "/Users/erjia/workspace/pytorch/torch/utils/data/dataloader.py", line 1200, in _next_data
    return self._process_data(data)
  File "/Users/erjia/workspace/pytorch/torch/utils/data/dataloader.py", line 1226, in _process_data
    data.reraise()
  File "/Users/erjia/workspace/pytorch/torch/_utils.py", line 420, in reraise
    raise self.exc
Exception: Re-raise the exception in DataLoader worker process 0
```

[ghstack-poisoned]

ejguan committed:
Update on "Add RemoteTraceback for exception with multiple arguments"


Fixes #25522 #30147 #34130 #44631 #48709 #51663

Taking reference of #40437 and [python](https://github.com/python/cpython/blob/e42b705188271da108de42b55d9344642170aa2b/Lib/concurrent/futures/process.py#L114-L131)

Due to lots of issues related to re-raising error, add a RemoteTraceback as the `__cause__` of the Exception in the main process.

When Python load object from different process using [pickle](https://docs.python.org/3/library/pickle.html#pickling-class-instances), it creates a new object with arguments. As the customized `Exception` is required for multiple arguments, which are not provided, it will causes the error.

```py
import torch.utils.data

class MyException(Exception):
    def __init__(self, msg, arg1, arg2):
        super().__init__(msg)

class MyDataset(torch.utils.data.IterableDataset):
    def __iter__(self):
        raise MyException("MyError", 1, 2)

if __name__ == '__main__':
    dl = torch.utils.data.DataLoader(MyDataset(), num_workers=1)
    list(iter(dl))
```

```
torch._utils._RemoteTraceback: Traceback (most recent call last):
  File "/Users/erjia/workspace/pytorch/torch/utils/data/_utils/worker.py", line 152, in _worker_loop
    fetcher = _DatasetKind.create_fetcher(dataset_kind, dataset, auto_collation, collate_fn, drop_last)
  File "/Users/erjia/workspace/pytorch/torch/utils/data/dataloader.py", line 52, in create_fetcher
    return _utils.fetch._IterableDatasetFetcher(dataset, auto_collation, collate_fn, drop_last)
  File "/Users/erjia/workspace/pytorch/torch/utils/data/_utils/fetch.py", line 21, in __init__
    self.dataset_iter = iter(dataset)
  File "/Users/erjia/workspace/test/test_exception.py", line 14, in __iter__
    raise MyException("MyError", 1, 2)
__mp_main__.MyException: MyError


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test/test_exception.py", line 19, in <module>
    list(iter(dl))
  File "/Users/erjia/workspace/pytorch/torch/utils/data/dataloader.py", line 518, in __next__
    data = self._next_data()
  File "/Users/erjia/workspace/pytorch/torch/utils/data/dataloader.py", line 1200, in _next_data
    return self._process_data(data)
  File "/Users/erjia/workspace/pytorch/torch/utils/data/dataloader.py", line 1226, in _process_data
    data.reraise()
  File "/Users/erjia/workspace/pytorch/torch/_utils.py", line 420, in reraise
    raise self.exc
Exception: Re-raise the exception in DataLoader worker process 0
```

[ghstack-poisoned]

ejguan commented on torch/_utils.py:
```diff
@@ -406,26 +414,21 @@ def __init__(self, exc_info=None, where="in background"):
 
     def reraise(self):
         r"""Reraises the wrapped exception in the current thread"""
-        # Format a message such as: "Caught ValueError in DataLoader worker
-        # process 2. Original Traceback:", followed by the traceback.
-        msg = "Caught {} {}.\nOriginal {}".format(
-            self.exc_type.__name__, self.where, self.exc_msg)
-        if self.exc_type == KeyError:
-            # KeyError calls repr() on its argument (usually a dict key). This
-            # makes stack traces unreadable. It will not be changed in Python
-            # (https://bugs.python.org/issue2651), so we work around it.
-            msg = KeyErrorMessage(msg)
-        elif getattr(self.exc_type, "message", None):
-            # Some exceptions have first argument as non-str but explicitly
-            # have message field
-            raise self.exc_type(message=msg)
-        raise self.exc_type(msg)
```
It's also doable to keep these here, and try:
```py
def reraise(self):
    ...
    self.exc = None
    try:
        self.exc = self.exc_type(msg)
    else:
         self.exc =  LocalException(....)
         self.exc.__cause__ = _RemoteTraceback
    raise self.exc
```
